0-common/igluctl/src/main/scala/com.snowplowanalytics.iglu/ctl/GenerateCommand.scala

```diff
@@ -119,7 +122,65 @@ case class GenerateCommand(
       outputPair.map(_._1),
       migrations,
       outputPair.flatMap(_._2),
-      warnings = schemaErrors ++ ddlErrors ++ ddlWarnings)
+      warnings = schemaVerWarnings ++ schemaVerErrors ++ schemaErrors ++ ddlErrors ++ ddlWarnings)
+  }
+
+
+  /**
+    * Checks if there is any missing schema version in a directory of schemas
+    * or if a specific schema file doesn't have version 1-0-0
+    *
+    * @param schemas list of valid JSON Schemas including all Self-describing information
+    * @return (versionWarnings, versionErrors)
+    */
+  private[ctl] def validateSchemaVersions(schemas: List[IgluSchema]): (List[String], List[String]) = {
+
+    def existMissingSchemaVersion(schemaMaps: List[SchemaMap]): Boolean = {
+      val numOfMaps = schemaMaps.length
+
+      if (numOfMaps == 1){
+        false
+      } else {
+        val prevModel    = schemaMaps.head.version.model
```

Comment:
The implementation itself can also be little bit generalized and simplified (later statement might sound questionable for you now, but will be clearer when we'll have a chance to talk Iglu ASTs) using two following facts:

1. `SchemaVer` is an ordered set (someone may argue that it's a [poset](https://en.wikipedia.org/wiki/Partially_ordered_set), but let's omit it for a moment). It means that we can build a `List` of `SchemaVer` from "lowest" to "highest" by `(versions: List[SchemaVer]).sorted) `(instance for `Ordering` is in iglu-core)
2. We can define very general (so general that they can live in `scala-core`) functions:

```scala
def modelSucc(a: SchemaVer.Full, b: SchemaVer.Full): Boolean = a.model + 1 == b.model
def revisionSucc(a: SchemaVer.Full, b: SchemaVer.Full): Boolean = a.revision + 1 == b.revision
def additionSucc(a: SchemaVer.Full, b: SchemaVer.Full): Boolean = a.addition + 1 == b.addition
def succ(a: SchemaVer.Full, b: SchemaVer.Full): Boolean = modelSucc(a, b) || revisionSucc(a, b) || additionSucc(a, b)
```

These functions tell us if `a` in some way is successor of `b` ([`succ`](http://hackage.haskell.org/package/base-4.10.1.0/docs/Prelude.html#v:succ) originates from Haskell).

Having these two facts we can write very simple function checking if in row of **ordered** `SchemaVer` sneaked a gap.



```scala
case class Acc(previous: Option[Full], gaps: List[SchemaVer])
object Acc {
  def empty: Acc = Acc(None, Nil)
}

orderedSchemaVers.foldLeft(Acc.empty) { (acc, cur) =>
  acc match {
    case Acc(None, gaps) if cur == Full(1,0,0) => Acc(Some(Full(1,0,0)), gaps)
    case Acc(None, gaps) => Acc(Some(cur), Full(1,0,0) :: gaps)
    case Acc(Some(prev), gaps) if succ(prev, cur) => Acc(Some(cur), gaps)
    case Acc(Some(prev), gaps) => Acc(Some(cur), cur :: gaps)
  }
}
```

This is quick and dirty implementation (not even entirely correct - it saves not "gaps", but latest found vers, up to you to fix it), but idea should be clear. I want this general functions to live somewhere is upstream libraries (Scala Core or Schema DDL) because this is a basis for a consistent Iglu repositories, where each schema does not live in isolation, but has a "connection" to its successors and predecessor and therefore basis for a schema inference.

Reply:
I'm not saying you need to implement it in R8, but this is a main reason why I was so excited about #221. If you're not going to implement this solution in R8 - then please create a ticket with link (or cite) to above message.

Reply:
Opened #315 for this feedback. I really liked the ideas, thanks for the feedback @chuwy !